### ISSUE TITLE
Potential alignment UB for 0-size slices in `slice_cast`

### ISSUE THREAD

**Issue by jsgf** [facebookexperimental/starlark-rust#9]
In `slice_cast`, if the slice has 0 size then it directly casts the pointer type and constructs a new zero-len slice:
https://github.com/facebookexperimental/starlark-rust/blob/3bb20164a18f1e0fe65ba9cfc4c5bbfc7f41f621/starlark/src/values/layout/arena.rs#L76-L78

`slice::from_raw_parts` requires [the pointer to be non-null and properly aligned](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety), so if `U` does not have the same alignment as `T` then this is UB:
> `data` must be non-null and aligned even for zero-length slices

I think the alignment assertion below should apply to the whole function, not just the zero-length case.

This function is also a bit odd in that it loses the length of a slice of zero-sized objects. For example, a non-zero length slice of `[()]` has a zero `mem::size_of_val`, so this would turn it into a zero-length `[U]`, rather than retaining the length (assuming `U` also has a zero size). I'm not sure if this is the intention; I guess it's OK if you're never expecting `T` to be zero-sized.

**Comment by ndmitchell** [comment]
Thanks for finding that. We never use ZST, so I've added an assertion to ensure that (easier to not think about them than write something for them). I've also removed the special case for empty slices - I don't think it added anything (I think it might have been designed for ZST support?), so now the assertion for alignment applies everywhere. Patch will land in this repo sometime in the next day or so.
